src/main/java/org/spongepowered/api/util/raytrace/RayTracingUtil.java

```diff
@@ -0,0 +1,145 @@
+     * @param location The location to start
+     * @param direction The ray direction
+     * @param consumer The consumer that consumes all visited blocks
+     */
+    public static void trace(Extent world, Vector3d location, Vector3d direction, BlockConsumer consumer) {
+        trace(location, direction, new PostionToBlockConsumer(world, BLOCK_LIMITER.apply(consumer)));
+    }
+
+    /**
+     * Starts the ray tracing from the given location.
+     *
+     * <p>Note: The range may be truncated by server settings.</p>
+     *
+     * @param location The location to start
+     * @param direction The ray direction
+     * @param consumer The consumer that consumes all visited blocks
+     */
+    public static void trace(Vector3d location, Vector3d direction, PositionConsumer consumer) {
+        double posX = location.getX();
+        double posY = location.getY();
+        double posZ = location.getZ();
+        // Does not move?
+        if (direction.lengthSquared() == 0) {
+            consumer.apply(posX, posY, posZ, direction);
+            return;
+        }
+        final Vector3d normalizedDirection = direction.normalize();
+        final double dx = normalizedDirection.getX();
+        final double dy = normalizedDirection.getY();
+        final double dz = normalizedDirection.getZ();
+        final double[] distances = new double[3];
+        final PositionConsumer limitedConsumer = POSITION_LIMITER.apply(consumer);
+        while (limitedConsumer.apply(posX, posY, posZ, direction)) {
+            distances[0] = distance(posX, dx);
+            distances[1] = distance(posY, dy);
+            distances[2] = distance(posZ, dz);
+            final double min = min(distances);
+            posX = sanitize(posX + min * dx);
+            posY = sanitize(posY + min * dy);
+            posZ = sanitize(posZ + min * dz);
+        }
+    }
+
+    // The minimum distance
+    private static double min(double[] moves) {
+        double min = moves[0];
+        if (min > moves[1]) {
+            min = moves[1];
+        }
+        if (min > moves[2]) {
+            min = moves[2];
+        }
+        return min;
+    }
+
+    // times how often you have to apply the given direction to pass any borders
+    private static double distance(double position, double direction) {
+        if (direction == 0) {
+            return Double.MAX_VALUE;
+        }
+        final double distanceLeft = (position % 1 + 1) % 1;
+        final double absDirection = Math.abs(direction);
+        if (distanceLeft == 0) {
+            return 1 / absDirection;
+        } else if (direction < 0) {
+            return distanceLeft / absDirection;
+        } else {
+            return (1 - distanceLeft) / absDirection;
+        }
+    }
+
+    // Sanitize the location (remove double fragments)
+    private static double sanitize(double d) {
+        final double dd = Math.abs(d % 1);
+        if ((dd != 0 && dd < 0.0000001) || (dd != 1 && dd > 0.9999999)) {
```

Comment:
The field is constant, so it will probably be pre-computed automatically. It does depend on the compiler/JVM.



> On May 16, 2015, at 8:04 PM, ST-DDT <notifications@github.com> wrote:
> 
> In src/main/java/org/spongepowered/api/util/raytrace/RayTracingUtil.java:
> 
> > +        }
> > +        final double distanceLeft = (position % 1 + 1) % 1;
> > +        final double absDirection = Math.abs(direction);
> > +        if (distanceLeft == 0) {
> > +            return 1 / absDirection;
> > +        } else if (direction < 0) {
> > +            return distanceLeft / absDirection;
> > +        } else {
> > +            return (1 - distanceLeft) / absDirection;
> > +        }
> > +    }
> > +
> > +    // Sanitize the location (remove double fragments)
> > +    private static double sanitize(double d) {
> > +        final double dd = Math.abs(d % 1);
> > +        if ((dd != 0 && dd < 0.0000001) || (dd != 1 && dd > 0.9999999)) {
> After some testing i need more than 1 * GenericMath.DBL_EPSILON, 2 *  GenericMath.DBL_EPSILON works perfectly fine.
> Shall i use my own constant for that or shall i let him calculate it every run?
> 
> if ((dd != 0 && dd < GenericMath.DBL_EPSILON * 2) || (dd != 1 && dd > 1 - GenericMath.DBL_EPSILON * 2)) {
> —
> Reply to this email directly or view it on GitHub.
>